### ISSUE TITLE
New update breaks ValheimPlus / server not updating to 0.148.7

### ISSUE THREAD

**Issue by Sloth-on-meth** [lloesche/valheim-server-docker#259]
Hi there!

ValheimPlus broke. when will the image be updated?

**Comment by gnouts** [comment]
I think it's update with your daily cron update :
https://github.com/lloesche/valheim-server-docker#updates-1
All users need to be disconnected.

Also, I think you can just restart your server and it will grab the latest patch from valheim and valheimplus

**Comment by lloesche** [comment]
> ValheimPlus broke

What about ValheimPlus is broken? Do you have any technical details?

**Comment by lloesche** [comment]
@gnouts the update check runs every 15 minutes. So there should be almost no lag between a new Valheim or ValheimPlus release and the server running it. By default it'll only update while there are no players connected.

**Comment by tz20** [comment]
So I'm having some trouble with getting the container to update as well. I've restarted it multiple times and downloaded the latest image. I also have Valheim Plus enabled.

Here's the message I see that suggests something is up: `valheim_1  | Mar 23 18:13:19 /supervisord: valheim-server Version generated with enforced mod : 0.147.3@0.9.5.5`

Also, when I connect I get the "Incompatible version" message and when I look at the logs I see this: `valheim_1  | Mar 23 18:26:09 /supervisord: valheim-server [Info   : Unity Log] 03/23/2021 18:26:09: VERSION check their:0.148.6@0.9.6 mine:0.147.3@0.9.5.5`

I'm running this locally on Windows and starting up the container with Docker Compose.

**Comment by TomRichter** [comment]
I also had this issue.  The logs showed Steam validation completing successfully, and that the ValheimPlus archive wasn't different so updating not needed.  But both Valheim and ValheimPlus showed the previous versions for the server when attempting to connect.

Failed attempts to fix include:
 * restarting the server and waiting to see the update messages
 * rebuilding the container
 * deleting the image and rebuilding the container

The real fix was stopping the container, deleting the contents of the `/opt/valheim` volume from the host (where Steam downloads the server binaries), then starting the container and letting it download again.  Since config files + world saves are in the `/config` volume, no progress was lost.

If it helps, I'm also running Docker in Windows Server 2022.

**Comment by lloesche** [comment]
So this happens when your container gets into an inconsistent state. Meaning it has downloaded the updated ValheimPlus ZIP archive but for whatever reason failed to extract and install it. Reason could be that the container was shut down while it was updating or if one was to manually modified the files in /opt/valheim or changed their permissions or install additional mod files, etc. What then happens is on next update check it sees that the local ZIP file is the most up to date one but fails to see that it hasn't been installed yet.

The way to get out of these situations quickly is to run the container once(!) with `DEBUG_START_FRESH=true`. This will wipe the `/opt/valheim/` data and recreate from scratch.

I also just pushed a commit that introduces a new flag `DEBUG_REINSTALL_VALHEIM_PLUS=true` that will only wipe and redownload the 6 MB V+ mod instead of the whole 1 GB Valheim server. Also moved some code around to make the install process more resilient to unexpected interruptions.

**Comment by tz20** [comment]
Yeah, I'm pretty sure I killed it on startup because I found some configs I wanted to change right after starting it. Thanks a ton for the help and the fix! I have been super happy with the image, the valheim plus support, and the extensive documentation!

**Comment by Sloth-on-meth** [comment]
@lloesche  i accidentally have 2 instances of vhplus somehow. if i do the reinstall command for the entire thing will my world be wiped?

**Comment by lloesche** [comment]
@Sloth-on-meth no, your worlds are in `/config` the flag only wipes `/opt/valheim`.

**Comment by TomRichter** [comment]
I'm running two containers attached to the same server data volume.  Should I have a separate volume per container/game server?

Could it be that one container started updates before the other and caused the patch to fail?  Although that still doesn't make sense for Steam verification failing -- I'd expect at least the main game version to have updated, especially when testing with only one container online.

**Comment by lloesche** [comment]
@TomRichter You should definitely have a separate volume per container. The update check runs every 15 minutes via cron meaning on updates you are guaranteed to have conflicting writes to the same files scheduled via the same clock at exactly the same time. Best case the two Steam instances just update the same files twice. Worst case it'll write garbage to the files and bring them into an inconsistent state. Or maybe it has some fencing mechanism, sees the other process accessing the files and be hands off. Though that's pretty unlikely, like why would they write code for two or more Steam instances running in different process namespaces updating the same game. Bottom line unless you are working for Valve and know the internals of Steams update process I'd just go with separate volumes per container.

Other than that I agree, with only one container online once you run the update process it should have corrected any potentially corrupt files via the `validate` flag. And because of the rsync mechanism the container is using for updates even a changed file that just got repaired for the same version would trigger a reinstall of V+.

Can you start the container and attach the entire startup log here?

**Comment by TomRichter** [comment]
Thanks for the suggestions.  I separated the volumes per container since disk space doesn't matter to me.  Hopefully that fixes patch day issues.

Sadly, I can't run the busted container because it's long since deleted.  I was just spitballing ideas to help anyone who might come after.

**Comment by lloesche** [comment]
@Sloth-on-meth did the flag resolve your issue?

**Comment by lloesche** [comment]
Closing for inactivity. Reopen if issue still persists.

**Comment by TomRichter** [comment]
The server files failed to update again from 0.148.6 to 0.148.7, despite the above changes.  This time, I kept a separate copy of the container + volumes and captured a clean start log + log of the last few days (which hopefully contains the patch release).  I'm not sure how to interpret these correctly, since it seems to say Steam validation went fine.  Any thoughts?

 * [clean_start.txt](https://github.com/lloesche/valheim-server-docker/files/6226470/clean_start.txt)
 * [long_log.txt](https://github.com/lloesche/valheim-server-docker/files/6226471/long_log.txt)

**Comment by InB4DevOps** [comment]
Had the same problem.
I stopped the server then deleted the content of the folder I mount to /opt/valheim and restarted the server. Only then did it  update to 0.148.7.

**Comment by InB4DevOps** [comment]
Re-opening since this still seems to be a problem.

**Comment by lloesche** [comment]
I'm afraid you are pretty much on your own debugging this issue as I am unable to reproduce it on any of our servers. I just checked our oldest server which has been running for weeks and it updated just fine.

If you read the attached `long_log.txt` you will see an error message:
```
Mar 29 12:16:34 /supervisord: valheim-updater mv: cannot move '/opt/valheim/plus' to '/opt/valheim/plus.old': No such file or directory
```
right after it downloaded the update.

The code responsible for moving the directory looks like this:
```
vp_install_path=/opt/valheim/plus               # ValheimPlus installation directory
...
...
    if [ -d "$vp_install_path" ]; then
        mv -f "$vp_install_path" "$vp_install_path.old"
    fi
```

Meaning in between the script checking whether `/opt/valheim/plus` exists (`if [ -d`) and it trying to move the directory to `.old` (`mv -f`) the directory was removed?!

As if you had multiple containers all accessing the same data volume or some other storage related conflict.


If I compare it to our server's logs it should look something like this
```
Mar 29 14:00:04 /supervisord: valheim-updater DEBUG - [33] - Received signal to check for update
Mar 29 14:00:04 /supervisord: valheim-updater DEBUG - [33] - Found CPU with 1200.869 MHz
Mar 29 14:00:04 /supervisord: valheim-updater DEBUG - [33] - Memory total/free/available: 65746572/860496/60600996
Mar 29 14:00:04 /supervisord: valheim-updater DEBUG - [33] - No players connected to Valheim server
Mar 29 14:00:04 /supervisord: valheim-updater INFO - Downloading/updating/validating Valheim server from Steam
...
...
Mar 29 14:00:15 /supervisord: valheim-updater INFO - Valheim Server was updated - restarting
Mar 29 14:00:15 /supervisord: valheim-updater DEBUG - [33] - ValheimPlus is enabled - running updater
Mar 29 14:00:15 /supervisord: valheim-updater INFO - Valheim Server was updated from Steam - signaling ValheimPlus updater to merge updated files
Mar 29 14:00:15 /supervisord: valheim-updater DEBUG - [14966] - Local ValheimPlus archive is identical to remote archive - no update required
Mar 29 14:00:15 /supervisord: valheim-updater INFO - Valheim dedicated server or ValheimPlus got updated - merging installation files
Mar 29 14:00:15 /supervisord: valheim-updater DEBUG - [14966] - Merging Valheim server and ValheimPlus mod
...
...
Mar 29 14:00:19 /supervisord: valheim-updater DEBUG - [14966] - Writing file to restart Valheim server
2021-03-29 14:00:20,299 INFO waiting for valheim-server to stop
Mar 29 14:00:20 /supervisord: valheim-server DEBUG - [29136] - Received signal to shut down valheim-server
Mar 29 14:00:20 /supervisord: valheim-server INFO - Shutting down Valheim server with PID 29140
...
...
Mar 29 14:00:26 /supervisord: valheim-server INFO - Shutdown complete
2021-03-29 14:00:26,311 INFO stopped: valheim-server (exit status 0)
2021-03-29 14:00:26,315 INFO spawned: 'valheim-server' with pid 15026
Mar 29 14:00:26 /supervisord: valheim-server INFO - Running Valheim Server
...
...
Mar 31 13:47:19 /supervisord: valheim-server 03/31/2021 13:47:19: VERSION check their:0.148.7@0.9.6  mine:0.148.7@0.9.6
Mar 31 13:47:19 /supervisord: valheim-server Version generator started.
Mar 31 13:47:19 /supervisord: valheim-server Version generated with enforced mod : 0.148.7@0.9.6
Mar 31 13:47:19 /supervisord: valheim-server Version generator started.
Mar 31 13:47:19 /supervisord: valheim-server Version generated with enforced mod : 0.148.7@0.9.6
Mar 31 13:47:19 /supervisord: valheim-server 03/31/2021 13:47:19: Server: New peer connected,sending global keys
Mar 31 13:47:22 /supervisord: valheim-server 03/31/2021 13:47:22: VPlus configuration synced to peer #1012455497
Mar 31 13:47:39 /supervisord: valheim-server 03/31/2021 13:47:39: Got character ZDOID from Luu : 1012455497:1
```

**Comment by TomRichter** [comment]
Thanks for the tips + partial fix.  I don't mind kicking it each time, but wanted to give logs since I couldn't before.  The two containers definitely have their own volumes, so I guess my plan ATM is hope it magically doesn't happen again.

I'm still glad this project exists and will continue using it, so thanks again for simplifying hosting <3
